cpp/src/converters/COOtoCSR.cuh

```diff
@@ -222,3 +225,121 @@ void ConvertCOOtoCSR_weighted(T const * sources, T const * destinations, W const
+VT sort(GraphCOOView<VT,ET,WT> &graph, cudaStream_t stream) {
+  VT max_src_id;
+  VT max_dst_id;
+  if (graph.has_data()) {
+    thrust::stable_sort_by_key(rmm::exec_policy(stream)->on(stream),
+        graph.dst_indices,
+        graph.dst_indices + graph.number_of_edges,
+        thrust::make_zip_iterator(thrust::make_tuple(graph.src_indices, graph.edge_data)));
+    CUDA_TRY(cudaMemcpy(&max_dst_id,
+          &(graph.dst_indices[graph.number_of_edges-1]),
+          sizeof(VT), cudaMemcpyDefault));
+    thrust::stable_sort_by_key(rmm::exec_policy(stream)->on(stream),
+        graph.src_indices,
+        graph.src_indices + graph.number_of_edges,
+        thrust::make_zip_iterator(thrust::make_tuple(graph.dst_indices, graph.edge_data)));
+    CUDA_TRY(cudaMemcpy(&max_src_id,
+          &(graph.src_indices[graph.number_of_edges-1]),
+          sizeof(VT), cudaMemcpyDefault));
+  } else {
+    thrust::stable_sort_by_key(rmm::exec_policy(stream)->on(stream),
+        graph.dst_indices,
+        graph.dst_indices + graph.number_of_edges,
+        graph.src_indices);
+    CUDA_TRY(cudaMemcpy(&max_dst_id,
+          &(graph.dst_indices[graph.number_of_edges-1]),
+          sizeof(VT), cudaMemcpyDefault));
+    thrust::stable_sort_by_key(rmm::exec_policy(stream)->on(stream),
+        graph.src_indices,
+        graph.src_indices + graph.number_of_edges,
+        graph.dst_indices);
+    CUDA_TRY(cudaMemcpy(&max_src_id,
+          &(graph.src_indices[graph.number_of_edges-1]),
+          sizeof(VT), cudaMemcpyDefault));
+  }
+  return std::max(max_src_id, max_dst_id) + 1;
+}
+
+template <typename VT, typename ET>
+rmm::device_buffer create_offset(
+    VT * source,
+    VT number_of_vertices,
+    ET number_of_edges,
+    cudaStream_t stream) {
+  //Offset array needs an extra element at the end to contain the ending offsets
+  //of the last vertex
+  rmm::device_buffer offsets_buffer(sizeof(ET)*(number_of_vertices+1), stream);
+  ET * offsets = static_cast<ET*>(offsets_buffer.data());
+
+  thrust::fill(rmm::exec_policy(stream)->on(stream),
+      offsets, offsets + number_of_vertices + 1, number_of_edges);
+  thrust::for_each(rmm::exec_policy(stream)->on(stream),
+      thrust::make_counting_iterator<ET>(1),
+      thrust::make_counting_iterator<ET>(number_of_edges),
+      [source, offsets]
+      __device__ (ET index) {
+        VT id = source[index];
+        if (id != source[index-1]) {
+          offsets[id] = index;
+        }
+      });
+  ET zero = 0;
+  CUDA_TRY(cudaMemcpy(offsets, &zero, sizeof(ET), cudaMemcpyDefault));
+  auto iter = thrust::make_reverse_iterator(offsets + number_of_vertices);
+  thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+      iter, iter + number_of_vertices + 1, iter, thrust::minimum<ET>());
+  return offsets_buffer;
+}
+
+} //namespace detail
+
+template <typename VT, typename ET, typename WT>
+std::unique_ptr<GraphCSR<VT ,ET ,WT>> coo_to_csr(GraphCOOView<VT, ET, WT> const &graph) {
```

Comment:
So is the intention that this implementation would replace ConvertCOOtoCSR and ConvertCOOtoCSR_Weighted in all places in the code?

Reply:
Yes, as far as I can tell these are the only two functions that actually allocate memory for csr information. That's why it made sense to me to tackle this here along with the graph objects.